src/include/denc.h

```diff
@@ -77,37 +77,69 @@ inline constexpr bool denc_supported = denc_traits<T>::supported;
 # include <sys/types.h>
 # include <sys/stat.h>
 # include <fcntl.h>
+
 # define ENCODE_STR(x) #x
 # define ENCODE_STRINGIFY(x) ENCODE_STR(x)
-# define DENC_DUMP_PRE(Type)			\
-  char *__denc_dump_pre = p.get_pos();
-  // this hackery with bits below is just to get a semi-reasonable
-  // distribution across time.  it is somewhat exponential but not
-  // quite.
-# define DENC_DUMP_POST(Type)			\
-  do {									\
-    static int i = 0;							\
-    i++;								\
-    int bits = 0;							\
-    for (unsigned t = i; t; bits++)					\
-      t &= t - 1;							\
-    if (bits > 2)							\
-      break;								\
-    char fn[PATH_MAX];							\
-    ::snprintf(fn, sizeof(fn),						\
-	     ENCODE_STRINGIFY(ENCODE_DUMP_PATH) "/%s__%d.%x", #Type,		\
-	     getpid(), i++);						\
-    int fd = ::open(fn, O_WRONLY|O_TRUNC|O_CREAT|O_CLOEXEC, 0644);		\
-    if (fd >= 0) {							\
-      size_t len = p.get_pos() - __denc_dump_pre;			\
-      int r = ::write(fd, __denc_dump_pre, len);			\
-      (void)r;								\
-      ::close(fd);							\
-    }									\
-  } while (0)
+
+template<typename T>
+class DencDumper {
+public:
+  DencDumper(const char* name,
+	     ceph::bufferlist::contiguous_appender& appender)
+    : name{name},
+      appender{appender},
+      offset{appender.get_logical_offset()},
+      start{appender.get_pos()}
+  {}
+  ~DencDumper() {
+    if (do_sample()) {
+      dump();
+    }
+  }
+private:
+  static bool do_sample() {
+    // this hackery with bits below is just to get a semi-reasonable
+    // distribution across time.  it is somewhat exponential but not
+    // quite.
+    i++;
+    int bits = 0;
+    for (unsigned t = i; t; bits++)
+      t &= t - 1;
+    return bits <= 2;
+  }
+  void dump() {
+    char fn[PATH_MAX];
+    ::snprintf(fn, sizeof(fn),
+	       ENCODE_STRINGIFY(ENCODE_DUMP_PATH) "/%s__%d.%x", name,
+	       getpid(), i++);
+    int fd = ::open(fn, O_WRONLY|O_TRUNC|O_CREAT|O_CLOEXEC, 0644);
+    if (fd < 0) {
+      return;
+    }
```

Comment:
We could use Ceph's `scope_guard` to prevent leak on exception:
```cpp
  auto close_fd = make_scope_guard([fd] {
    ::close(fd);
  });
```
IMO this a nit, not a blocker.